### ISSUE TITLE
The test suite doesn't run on latest OSX

### ISSUE THREAD

**Issue by JulienPeloton** [ninia/jep#113]
- OS Platform, Distribution, and Version: OSX, High Sierra
- Python Distribution and Version: python3.6 (anaconda)
- Java Distribution and Version: JDK 1.8

Hi,

I downloaded jep, built it and installed it (python setup.py build install). 
Then if I launch the test suite, I got the following error:

```
$ python setup.py test
numpy include found at /Users/julien/anaconda3/lib/python3.6/site-packages/numpy/core/include
running test
running build
running setup_java
running build_java
running build_javah
running build_jar
running build_py
running build_ext
running build_scripts
ln -sf jep.cpython-36m-darwin.so build/lib.macosx-10.7-x86_64-3.6/libjep.jnilib
java.lang.UnsatisfiedLinkError: /Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so: dlopen(/Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so, 1): Library not loaded: @rpath/libpython3.6m.dylib
  Referenced from: /Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so
  Reason: image not found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1861)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at jep.Jep.<clinit>(Jep.java:217)
	at jep.Run.run(Run.java:52)
	at jep.Run.main(Run.java:151)
error: Unit tests failed with exit status 1
```

I have seen in a few other issues the same kind of error (but without real solution), but here it happens during the test phase...

Note: if I install jep through pip, I get the same kind of error. Example

```
$ pip install jep
... [OK] ...
$ which jep
/Users/julien/anaconda3/bin/jep
$ jep
java.lang.UnsatisfiedLinkError: /Users/julien/anaconda3/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so: dlopen(/Users/julien/anaconda3/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so, 1): Library not loaded: @rpath/libpython3.6m.dylib
  Referenced from: /Users/julien/anaconda3/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so
  Reason: image not found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1861)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at jep.Jep.<clinit>(Jep.java:217)
	at jep.Run.run(Run.java:52)
	at jep.Run.main(Run.java:151)
```

Any ideas?
Thanks!

Julien

**Comment by JulienPeloton** [comment]
Hi,

Thanks for your quick reply. Yes the file exists, and there is a symbolic link as expected:
```
libjep.jnilib -> jep.cpython-36m-darwin.so
```

If I perform the test just after the build (after a fresh clean), I have the same error:
```
$ python setup build test
... [build OK] ...
numpy include found at /Users/julien/anaconda3/lib/python3.6/site-packages/numpy/core/include
running test
running build
running setup_java
running build_java
running build_javah
running build_jar
running build_py
running build_ext
running build_scripts
ln -sf jep.cpython-36m-darwin.so build/lib.macosx-10.7-x86_64-3.6/libjep.jnilib
java.lang.UnsatisfiedLinkError: /Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so: dlopen(/Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so, 1): Library not loaded: @rpath/libpython3.6m.dylib
  Referenced from: /Users/julien/Documents/workspace/otrepos/jep/build/lib.macosx-10.7-x86_64-3.6/jep.cpython-36m-darwin.so
  Reason: image not found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1861)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at jep.Jep.<clinit>(Jep.java:217)
	at jep.Run.run(Run.java:52)
	at jep.Run.main(Run.java:151)
error: Unit tests failed with exit status 1
```
In addition after the build, if I go to `build/script-3.6`, and execute jep I get:

```
$ ./jep
Error: Could not find or load main class jep.Run
```

Thanks!
Julien

**Comment by ndjensen** [comment]
Yeah, the jep command line script/interpreter won't work until it's installed due to some variable substitution that goes on in there.  After it's installed, can you open up the command line script in a text editor and see what `LD_LIBRARY_PATH` is.  Does `LD_LIBRARY_PATH` include the path to the directory containing `libpython3.6m.dylib`?  If not, it's worth a shot to modify that environment variable to include that directory and then run jep at the command line and see if that fixes it.

**Comment by JulienPeloton** [comment]
Hi,
Yes, `LD_LIBRARY_PATH` includes the path to `libpython3.6m.dylib`. I tried a few other things (like setting the `DYLD_LIBRARY_PATH` and `LD_PRELOAD`, or linking differently the library - but without success). I'm afraid this is again a conflict between anaconda and virtualenv (already discussed in some previous issues). 
In the meantime, I set up a virtual machine with Linux - and everything works fine. Not sure I want to investigate more on this (as it appears to be not-so-jep related).
Thanks for your time!

Julien

**Comment by ndjensen** [comment]
@jhegedus42, what version of numpy are you using?

**Comment by bsteffensmeier** [comment]
e5dde73 in dev_3.8 fixes the crash that @jhegedus42 is seeing. The crash is caused because numpy is not loading, and the fix only stops the crash, numpy will still fail to load which will result in test failures. I would guess that numpy is not loading because of incomplete or conflicting versions of numpy being found which can easily happen when you have multiple versions of python installed.

**Comment by bsteffensmeier** [comment]
I suspect this problem may be occurring because System Integrity Protection is preventing child processes from inheriting the DYLD_LIBRARY_PATH. I don't have a mac to experiment on. @canada11  would you be willing to try [disabling System Integrity Protection](https://developer.apple.com/library/content/documentation/Security/Conceptual/System_Integrity_Protection_Guide/ConfiguringSystemIntegrityProtection/ConfiguringSystemIntegrityProtection.html), setting  DYLD_LIBRARY_PATH, and seeing if it works?

**Comment by canada11** [comment]
What does the DYLD_LIBRARY_PATH need to be? (or where can I see it?)

**Comment by bsteffensmeier** [comment]
DYLD_LIBRARY_PATH should point to the directory that contains libpython3.6m.dylib.

**Comment by canada11** [comment]
Disabling the system integrity protection didn't seem to change anything, adding the DYLD_LIBRARY_PATH to my ~/.bash_profile did change the error a *little*, now it says this:
```
$jep
java.lang.UnsatisfiedLinkError: /anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib: dlopen(/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib, 1): Library not loaded: @rpath/libpython3.6m.dylib
  Referenced from: /anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib
  Reason: image not found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1861)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at jep.Jep.<clinit>(Jep.java:217)
	at jep.Run.run(Run.java:52)
	at jep.Run.main(Run.java:151)
```

**Comment by canada11** [comment]
Disabling the system integrity protection and adding the DYLD_LIBRARY_PATH to my ~/.bash_profile did change the error a *little*, it now says this:
```
$jep
java.lang.UnsatisfiedLinkError: /anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib: dlopen(/anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib, 1): Library not loaded: @rpath/libpython3.6m.dylib
  Referenced from: /anaconda3/lib/python3.6/site-packages/jep/libjep.jnilib
  Reason: no suitable image found.  Did find:
	/anaconda3/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so/libpython3.6m.dylib: stat() failed with errno=20
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1861)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at jep.Jep.<clinit>(Jep.java:217)
	at jep.Run.run(Run.java:52)
	at jep.Run.main(Run.java:151)
```
Of note: I use pip to install jep and it does not by default make a symlink of the libjep.jnilib file in the site-packages/jep/ location. If I make it a symlink to the jep.cpython-36m-darwin.so file in the same folder then it still has the above error.

**Comment by canada11** [comment]
It looks like it's probably as simple as getting one of the paths correct. It seems to just never reach the console.py file. As far as I can tell. I just can't figure out what is needing to point where to get it to run correctly.

**Comment by bsteffensmeier** [comment]
What did you set DYLD_LIBRARY_PATH to for each of those python installs? You might be able to get the correct setting if you start python normally and do the following:

```
>>> from distutils import sysconfig
>>> print(sysconfig.get_config_var('LIBDIR'))
```

**Comment by canada11** [comment]
I got it to work when I installed it from source code using `python setup.py build install`
The paths that DYLD_LIBRARY_PATH were set to were:
```
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so
```
and 
```
/anaconda3/lib/python3.6/site-packages/jep/jep.cpython-36m-darwin.so
```
I did not have virtualenv installed (at any point). I also had two python versions at any time, the system python which I try not to touch, and then either anaconda/brew installed python/or source code installed python, but only one at a time and it was always 3.6

I don't know why the install from source vs pip made a difference, but it fixed it. Hopefully this can help others!

**Comment by JulienPeloton** [comment]
@canada11  thanks for opening again this issue, and for your investigations.
I followed your recommendation (build from source), but still does not work for me.
Did you disabled the system integrity protection as well to get it to work?

Julien

**Comment by ndjensen** [comment]
@JulienPeloton, can you clarify or edit your previous comment?  Your code samples show LD_LIBRARY_PATH but your comment says "set DYLD_LIBRARY_PATH".

**Comment by JulienPeloton** [comment]
My bad - unfortunate copy/paste. ``LD_LIBRARY_PATH`` has to contain the python lib (corrected above as well).

Thanks for all the good work!

**Comment by ndjensen** [comment]
System Integrity Protection causes it to ignore `LD_LIBRARY_PATH` and `DYLD_LIBRARY_PATH`, hence why it doesn't work with it on.

If you have some time and/or want to try it with System Integrity Protection on, could you try getting Jep's dev_3.8 branch?  Then set the environment variable `PYTHONHOME=/Users/julien/anaconda3` since that appears to be your python home directory.  Test if it builds and the jep script runs without `LD_LIBRARY_PATH` set.  @bsteffensmeier did work on the 3.8 branch to try and reduce dependencies on `LD_LIBRARY_PATH`.

**Comment by ndjensen** [comment]
On dev_3.8 branch I added `PYTHONHOME` to the test command and to the jep script when running on OS X, so this should be resolved when Jep 3.8 releases.

**Comment by ndjensen** [comment]
3.8 is now released so closing this issue as fixed.
